File: Try_GameEngine5/src/com/example/try_gameengine/framework/ILayer.java
package com.example.try_gameengine.framework;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.ConcurrentLinkedDeque;
import java.util.concurrent.ConcurrentLinkedQueue;

import com.example.try_gameengine.framework.ALayer.LayerParam;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PointF;
import android.graphics.RectF;
import android.view.MotionEvent;

//import com.example.try_gameengine.framework.ILayer.OnLayerClickListener;
//import com.example.try_gameengine.framework.ILayer.OnLayerLongClickListener;

public interface ILayer extends Cloneable{
	
	public void setPosition(float x, float y);

	/** * 绘制自己的抽象接口 * * @param canvas * @param paint */
	public abstract void drawSelf(Canvas canvas, Paint paint);

	public RectF getSmallViewRect();

	public void setSmallViewRect(RectF smallViewRect);

	public void remove(ILayer layer);
	
	public boolean isAutoAdd();

	public void addWithLayerLevelIncrease(ILayer layer);
	public void addWithLayerLevelIncrease(ILayer layer, int increaseNum);

	public void addWithOutLayerLevelIncrease(ILayer layer);
	
	public void addWithLayerLevel(ILayer layer, int layerLevel);
	
	//composite
	public void addChild(ILayer layer);

	public ILayer getChild(int i);
	
	public List<ILayer> getLayers();

	public Iterator createIterator();

	public void moveAllChild(int offsetLayerLevel);
	
	public void setParent(ILayer parent);
	
	public ILayer getParent();
	
	public void setInitWidth(int w);
	
	public void setInitHeight(int h);
	
	public void setWidth(int w);
	
	public void setHeight(int h);
	
	public int getWidth();
	
	public int getHeight();
	
	public float getX();
	
	public float getLeft();
	
	public float getCenterX();
	
	public void setX(float x);
	
	public float getY();
	
	public float getTop();
	
	public float getCenterY();
	
	public void setY(float y);
	
	public void setBitmapAndAutoChangeWH(Bitmap bitmap);
	
	public void setBitmap(Bitmap bitmap);
	
	public Bitmap getBitmap();
	
	public RectF getDst();

	public int getLayerLevel();

	public void setLayerLevel(int layerLevel);

	public int getAlpha();

	public void setAlpha(int alpha);

	public Paint getPaint();

	public void setPaint(Paint paint);
	
	public void removeFromParent();
	
	public void removeFromAuto();

	public int getzPosition();

	public void setzPosition(int zPosition);

//	public boolean iszPositionValid();
	
	public boolean isTouching();

	public void setTouching(boolean isTouching);
	
	public boolean isPressed();
	
	public void setPressed(boolean pressed);

	public boolean isComposite();

	public void setComposite(boolean isComposite);
	
	public PointF getLocationInScene();

	public void setLocationInScene(PointF locationInScene);

	public PointF locationInLayer(float x, float y);
	
	public PointF locationInSceneByCompositeLocation(float locationInLayerX, float locationInLayerY);
	
	public RectF frameInSceneByCompositeLocation();
	
	public ILayer getRootLayer();
	
	public List<ILayer> getLayersFromRootLayerToCurrentLayerInComposite();
	
//	public void setOnLayerClickListener(OnLayerClickListener onLayerClickListener);
//	
//	public void setOnLayerLongClickListener(OnLayerLongClickListener onLayerLongClickListener);
	
	public boolean onTouchEvent(MotionEvent event);
	
	public Object clone() throws CloneNotSupportedException;

	public LayerParam getLayerParam();

	public RectF getFrame();
	
	public RectF getFrameInScene();
	
	public void setFrameInScene(RectF frameInScene);
	
	public boolean isClipOutside();
	
	public void setBackgroundColor(int backgroundColor);
	
	public void setFlag(int flag);
	
	public int getFlag();
	
	public void addFlag(int flag);
	
	public void removeFlag(int flag);
	
	public boolean isUsedzPosition();
	
	public void setAutoAdd(boolean autoAdd);
	
	public boolean isEnable();
	
	public void setEnable(boolean isEnable);
	
	public boolean isHidden();
	
	public void setHidden(boolean isHidden);
	
	public boolean isVisible();
	
	public void setVisible(boolean isVisible);
}

File: Try_GameEngine5/src/com/example/try_gameengine/framework/ILayer.kt
package com.example.try_gameengine.framework

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.Point
import android.graphics.PointF
import android.graphics.RectF
import android.view.MotionEvent
import com.example.try_gameengine.framework.ALayer.LayerParam

//import com.example.try_gameengine.framework.ILayer.OnLayerClickListener;
//import com.example.try_gameengine.framework.ILayer.OnLayerLongClickListener;
interface ILayer : Cloneable {
    fun setPosition(x: Float, y: Float)
    fun getPosition(): PointF?

    /** * 绘制自己的抽象接口 * * @param canvas * @param paint  */
    fun drawSelf(canvas: Canvas?, paint: Paint?)

    fun getSmallViewRect(): RectF?
    fun setSmallViewRect(smallViewRect: RectF?)

    fun remove(layer: ILayer?)

    fun isAutoAdd(): Boolean
    fun setAutoAdd(isAutoAdd: Boolean)

    fun addWithLayerLevelIncrease(layer: ILayer?)
    fun addWithLayerLevelIncrease(layer: ILayer?, increaseNum: Int)

    fun addWithOutLayerLevelIncrease(layer: ILayer?)

    fun addWithLayerLevel(layer: ILayer?, layerLevel: Int)

    //composite
    fun addChild(layer: ILayer?)

    fun getChildAt(index: Int): ILayer?

    fun getChildCount(): Int

    fun getLayers(): MutableList<ILayer>

    fun createIterator(): MutableIterator<*>?

    fun getParent(): ILayer?
    fun setParent(parent: ILayer?)

    fun setInitWidth(w: Int)

    fun setInitHeight(h: Int)

    fun getWidth(): Int
    fun setWidth(width: Int)

    fun getHeight(): Int
    fun setHeight(height: Int)

    fun setSize(w: Int, h: Int)

    fun getSize(): Point?

    fun getX(): Float
    fun setX(x: Float)

    fun getLeft(): Float

    fun getCenterX(): Float

    fun getY(): Float
    fun setY(y: Float)

    fun getTop(): Float

    fun getCenterY(): Float

    fun setBitmapAndAutoChangeWH(bitmap: Bitmap?)

    fun getBitmap(): Bitmap?
    fun setBitmap(bitmap: Bitmap?)

    fun getDst(): RectF

    fun getLayerLevel(): Int
    fun setLayerLevel(layerLevel: Int)

    fun getAlpha(): Int
    fun setAlpha(alpha: Int)

    fun getPaint(): Paint?
    fun setPaint(paint: Paint?)

    fun removeFromParent()

    fun removeFromAuto()

    fun getzPosition(): Int

    fun setzPosition(zPosition: Int)

    //	public boolean iszPositionValid();
    fun isTouching(): Boolean
    fun setTouching(isTouching: Boolean)

    fun isPressed(): Boolean
    fun setPressed(isPressed: Boolean)

    fun isComposite(): Boolean
    fun setComposite(isComposite: Boolean)

    fun getLocationInScene(): PointF?
    fun setLocationInScene(locationInScene: PointF?)

    fun locationInLayer(x: Float, y: Float): PointF?

    fun locationInSceneByCompositeLocation(
        locationInLayerX: Float,
        locationInLayerY: Float
    ): PointF?

    fun frameInSceneByCompositeLocation(): RectF?

    fun getRootLayer(): ILayer?

    fun getLayersFromRootLayerToCurrentLayerInComposite(): MutableList<ILayer>

    //	public void setOnLayerClickListener(OnLayerClickListener onLayerClickListener);
    //	
    //	public void setOnLayerLongClickListener(OnLayerLongClickListener onLayerLongClickListener);
    fun onTouchEvent(event: MotionEvent?): Boolean

    fun onTouchEvent(event: MotionEvent?, touchEventFlag: Int): Boolean

    fun calculateWHByChildern()

    fun isAutoSizeByChildren(): Boolean

    fun autoCalculateSizeByChildern(): RectF?

    @Throws(CloneNotSupportedException::class)
    public override fun clone(): Any

    fun getLayerParam(): LayerParam

    fun getFrame(): RectF

    fun getFrameInScene(): RectF
    fun setFrameInScene(frameInScene: RectF?)

    fun isClipOutside(): Boolean

    fun setBackgroundColor(backgroundColor: Int)

    fun getFlag(): Int
    fun setFlag(flag: Int)

    fun addFlag(flag: Int)

    fun removeFlag(flag: Int)

    fun isUsedzPosition(): Boolean

    fun isEnable(): Boolean
    fun setEnable(isEnable: Boolean)

    fun isHidden(): Boolean
    fun setHidden(isHidden: Boolean)

    fun isVisible(): Boolean
    fun setVisible(isVisible: Boolean)

    fun frameTrig()

    fun getLayerMatrix(): Matrix

    fun getClipedCanvas(canvas: Canvas?, paint: Paint?): Canvas?
}
